docs/client_middleware_cookbook.rst

```diff
@@ -0,0 +1,330 @@
+
+Basic Authentication Middleware
+-------------------------------
+
+Basic authentication is a simple authentication scheme built into the HTTP protocol.
+Here's a middleware that automatically adds Basic Auth headers to all requests:
+
+.. code-block:: python
+
+    import base64
+    from typing import Optional
+    from aiohttp import ClientRequest, ClientResponse, ClientHandlerType
+
+    class BasicAuthMiddleware:
+        """Middleware that adds Basic Authentication to all requests."""
+
+        def __init__(self, username: str, password: str):
+            self.username = username
+            self.password = password
+            self._auth_header = self._encode_credentials()
+
+        def _encode_credentials(self) -> str:
+            """Encode username and password to base64."""
+            credentials = f"{self.username}:{self.password}"
+            encoded = base64.b64encode(credentials.encode()).decode()
+            return f"Basic {encoded}"
+
+        async def __call__(
+            self,
+            request: ClientRequest,
+            handler: ClientHandlerType
+        ) -> ClientResponse:
+            """Add Basic Auth header to the request."""
+            # Only add auth if not already present
+            if "Authorization" not in request.headers:
+                request.headers["Authorization"] = self._auth_header
+
+            # Proceed with the request
+            return await handler(request)
+
+Usage example:
+
+.. code-block:: python
+
+    import aiohttp
+    import asyncio
+    import logging
+
+    _LOGGER = logging.getLogger(__name__)
+
+    async def main():
+        # Create middleware instance
+        auth_middleware = BasicAuthMiddleware("user", "pass")
+
+        # Use middleware in session
+        async with aiohttp.ClientSession(middlewares=(auth_middleware,)) as session:
+            async with session.get("https://httpbin.org/basic-auth/user/pass") as resp:
+                _LOGGER.debug("Status: %s", resp.status)
+                data = await resp.json()
+                _LOGGER.debug("Response: %s", data)
+
+    asyncio.run(main())
+
+.. _cookbook-retry-middleware:
+
+Simple Retry Middleware
+-----------------------
+
+A retry middleware that automatically retries failed requests with exponential backoff:
+
+.. code-block:: python
+
+    import asyncio
+    import logging
+    from typing import Optional, Set
+    from aiohttp import ClientRequest, ClientResponse, ClientHandlerType
+
+    _LOGGER = logging.getLogger(__name__)
+
+    class RetryMiddleware:
+        """Middleware that retries failed requests with exponential backoff."""
+
+        def __init__(
+            self,
+            max_retries: int = 3,
+            retry_statuses: Optional[Set[int]] = None,
+            initial_delay: float = 1.0,
+            backoff_factor: float = 2.0
+        ):
+            self.max_retries = max_retries
+            self.retry_statuses = retry_statuses or {429, 500, 502, 503, 504}
```

Comment:
Should these come via `from http import HTTPStatus`?

```suggestion
            self.retry_statuses = retry_statuses or {HTTPStatus.TOO_MANY_REQUESTS, HTTPStatus.INTERNAL_SERVER_ERROR, HTTPStatus.BAD_GATEWAY, HTTPStatus.SERVICE_UNAVAILABLE, HTTPStatus.GATEWAY_TIMEOUT}
```

Reply:
09079407e